refactor: EventLoop locking cleanups + client disconnect exception

ryanofsky opened the pull request:
This PR was originally written as a code cleanup to follow up on review comments in earlier PRs.  Several other commits were added afterward building on the earlier changes, and necessary for the bugfixes in https://github.com/bitcoin/bitcoin/pull/32345, which resolves issues #123, #176 and #182.

Summary of changes:

- The commit "Improve IPC client disconnected exceptions" is the only external change, and lets clients detect when they are trying to call a remote method after a connection has been closed. This change is used by
[bitcoin/bitcoin#32345](https://github.com/bitcoin/bitcoin/pull/32345) commit "ipc: Handle bitcoin-wallet disconnection" to fix [#123](https://github.com/bitcoin-core/libmultiprocess/issues/123).
- The commits "Prevent EventLoop async cleanup thread early exit during shutdown" and "Prevent IPC server crash if disconnected during IPC call" fix issues reported in [#182](https://github.com/bitcoin-core/libmultiprocess/issues/182).
- New tests are added in other commits to verify these fixes, covering different kinds of disconnections.
- The change to detect disconnects also relies on an earlier "Add ProxyContext EventLoop* member" commit.
- An "Add EventLoopRef RAII class" commit simplifies a recent bugfix in [#159](https://github.com/bitcoin-core/libmultiprocess/issues/159) and also relies on the earlier "EventLoop* member" commit.
- The "Remove DestructorCatcher and AsyncCallable" commit follows up on [#144 (comment)](https://github.com/chaincodelabs/libmultiprocess/pull/144#discussion_r1939618932).
- The "Add clang thread safety annotations" commit follows up on [#129 (comment)](https://github.com/bitcoin-core/libmultiprocess/pull/129#discussion_r1928455500).

vasild committed:
refactor: Remove DestructorCatcher and AsyncCallable

Use kj::Function instead of std::function to avoid the need for AsyncCallable
and DestructorCatcher classes, which were used to work around the requirement
that std::function objects need to be copyable. kj::Function does not have this
requirement.

Change is from https://github.com/chaincodelabs/libmultiprocess/pull/144#discussion_r1941178169

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ryanofsky committed:
proxy-io.h: Add EventLoopRef RAII class handle addClient/removeClient refcounting

ryanofsky committed:
proxy-io.h: Add more detailed EventLoop comment

ryanofsky committed:
refactor: Use EventLoopRef instead of addClient/removeClient

Use EventLoopRef to avoid reference counting bugs and be more exception safe

ryanofsky committed:
refactor: Add ProxyContext EventLoop* member

This commit makes mechanical changes needed to simplify an upcoming commit
which replaces EventLoop* with an EventLoopRef.

This change also happens to be also useful on its own so clientInvoke can
detect disconnections in a non-racy way
(https://github.com/bitcoin-core/libmultiprocess/issues/123#issuecomment-2516088243)
by seeing if the client Connection pointer is null while holding the event loop
mutex.

ryanofsky committed:
refactor: Drop addClient/removeClient methods

Now that they are only called in one place by EventLoopRef class, they can be
inlined.

ryanofsky commented:
Because this change removes `EventLoop` `addClient` and `removeClient` methods it will require an update to Bitcoin core if it is merged.

```diff
--- a/src/ipc/capnp/protocol.cpp
+++ b/src/ipc/capnp/protocol.cpp
@@ -41,10 +41,7 @@ class CapnpProtocol : public Protocol
 public:
     ~CapnpProtocol() noexcept(true)
     {
-        if (m_loop) {
-            std::unique_lock<std::mutex> lock(m_loop->m_mutex);
-            m_loop->removeClient(lock);
-        }
+        m_loop_ref.reset();
         if (m_loop_thread.joinable()) m_loop_thread.join();
         assert(!m_loop);
     };
@@ -83,10 +80,7 @@ public:
         m_loop_thread = std::thread([&] {
             util::ThreadRename("capnp-loop");
             m_loop.emplace(exe_name, &IpcLogFn, &m_context);
-            {
-                std::unique_lock<std::mutex> lock(m_loop->m_mutex);
-                m_loop->addClient(lock);
-            }
+            m_loop_ref.emplace(*m_loop);
             promise.set_value();
             m_loop->loop();
             m_loop.reset();
@@ -96,6 +90,7 @@ public:
     Context m_context;
     std::thread m_loop_thread;
     std::optional<mp::EventLoop> m_loop;
+    std::optional<mp::EventLoopRef> m_loop_ref;
 };
 } // namespace
```

ryanofsky committed:
Improve IPC client disconnected exceptions

Improve clientInvoke exceptions so IPC clients can more reliably detect when
they are calling a remote method after the connection is closed. Before this
change different exceptions were thrown, which made this condition difficult to
detect and handle.

ryanofsky committed:
test: Add test coverage for client & server disconnections

ryanofsky commented:
Updated ce4814f46d3c38d0bdd08df2253bc69b084f22ee -> b47ea9f69307908680595b96720bf41464bb3089 ([`pr/eventlock.1`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.1) -> [`pr/eventlock.2`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.2), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/eventlock.1..pr/eventlock.2)), make test and comment fixes, splitting commits, and adding connection disconnects tests and exceptions to help resolve #123

ryanofsky commented:
Rebased b47ea9f69307908680595b96720bf41464bb3089 -> f15ef6cdeec54c83f52e00de47d57b09f9a5f03b ([`pr/eventlock.2`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.2) -> [`pr/eventlock.3`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.3), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/eventlock.2-rebase..pr/eventlock.3)) so this can be merged cleanly in bitcoin PR https://github.com/bitcoin/bitcoin/pull/32345. Also dropped DisconnectError exception type so original ipc::Exception can be used instead.

ryanofsky committed:
refactor: Add clang thread safety annotations to EventLoop

Add basic thread safety annotations to EventLoop. Use could be expanded and
deepened but this puts basic functionality in place.

Use of annotations was discussed in
https://github.com/chaincodelabs/libmultiprocess/pull/129#discussion_r1928455500

ryanofsky commented:
Rebased f15ef6cdeec54c83f52e00de47d57b09f9a5f03b -> 22143588944af415bd420eb0418b331d1fd3cd86 ([`pr/eventlock.3`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.3) -> [`pr/eventlock.4`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.4), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/eventlock.3-rebase..pr/eventlock.4)) to fix conflict with #165

ryanofsky commented:
Rewrote the PR description since this is now a dependency of https://github.com/bitcoin/bitcoin/pull/32345 and needed to fix bugs #123 and #176

Sjors reviewed (approved):
ACK 6715a967d98c05a936a0ede153cc7c485d36590f

The changes here make sense to me, at least at a high level.

ryanofsky committed:
doc: Document ProxyClientBase destroy_connection option

Suggested by Sjors Provoost <sjors@sprovoost.nl>
https://github.com/bitcoin-core/libmultiprocess/pull/160#discussion_r2135564583

ryanofsky reviewed (commented):
Thanks for the review!

Added 1 commit 6715a967d98c05a936a0ede153cc7c485d36590f -> 0b45cb5662798553bc51979125d5c549d05e3262 ([`pr/eventlock.5`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.5) -> [`pr/eventlock.6`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.6), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/eventlock.5...pr/eventlock.6)) documenting ProxyClientBase destroy_connection option

ryanofsky committed:
Prevent EventLoop async cleanup thread early exit during shutdown

Antoine Poinsot <darosior@protonmail.com> reported a bug with details in
https://github.com/bitcoin-core/libmultiprocess/issues/182#issuecomment-2960280611
where an IPC client rapidly connecting and disconnecting to the server in a
loop could cause problems.

One problem, fixed by this commit, was that if a server process is shutting
down, the async cleanup thread in `EventLoop::startAsyncThread` responsible for
destroying unused server object on unclean disconnects could detect the done()
condition and decide to exit right before a new incoming connection is
processed, and exit prematurely when there might be more cleanup work for it
do. If this happens, process shutdown could hang waiting for cleanup work that
will never be completed.

This commit fixes that problem by changing the `EventLoop::startAsyncThread()`
while condition to check whether the `EventLoop::loop() method has _exited_,
instead of checking whether it is about to exit. Specifically the change makes
the `m_async_fns` list into a `std::optional` value and sets it to `nullopt`
when the `loop()` method exits.

ryanofsky committed:
test: Test disconnects during IPC calls

This change adds two unit tests to make sure shutdown happens cleanly if the
the client disconnects while the server is processing an IPC call.

The first test checks a synchronous case which has always worked, where the
disconnect starts but is not processed by the event loop until after the IPC
call completes.

The second test checks an asynchronous case which was fixed by the previous
commit, where the disconnect is processed before the IPC call generates a
response. This test would usually crash or hang prior to the fix in the
previous commit. It would crash reliably with address sanitizer.

ryanofsky committed:
Prevent IPC server crash if disconnected during IPC call

Antoine Poinsot <darosior@protonmail.com> reported a bug with details in
https://github.com/bitcoin-core/libmultiprocess/issues/182#issuecomment-2960280611
where an IPC client rapidly connecting and disconnecting to the server in a
loop could cause problems.

One problem this uncovered was hangs on shutdown fixed by the previous commit.

Another problem it uncovered is that if a disconnect happened while the IPC
server was executing an IPC call, and the onDisconnect() handling code ran
before the IPC call finished, memory corruption would occur usually leading to
a hang or crash.

This happend because the ~ProxyServerBase() destructor was written with the
assumption that it would always be called either from the ~Connection()
destructor, or before that point, so its m_context.connection pointer would
always be valid. Typically this was the case, but not if the connection was
closed during an active IPC call.

A unit test to reproduce this error is added in the subsequent commit.

The fix for the bug here is actually a code simplification.  Previously
~ProxyServerBase() destructor would append a std::function callback destroying
the ProxyServerBase::m_impl object to
m_context.connection->m_async_cleanup_fns, so the object destructor could be
called asynchronously without blocking the event loop. Now it just appends the
same callback function to m_context.loop->m_async_fns without going through the
Connection object.

The new code is an improvement over the previous code in two other ways:

- It is a code simplification since previous code was needlessly complicated.

- In the case where there is no disconnect, and the remote ProxyClient is
  destroyed, and no "destroy" method is declared, this change causes the
  ProxyServer::m_impl object to be freed shortly after the client is freed,
  instead of being delayed until the connection is closed. (If a "destroy"
  method is declared, this change has no effect because in that case destroying
  the ProxyClient calls destroy and destroys ProxyServer::m_impl synchronously.)

The previous code was trying to make the ProxyServer cleanup with
Connection::m_async_cleanup_fns mirror the ProxyClient cleanup with
Connection::m_sync_cleanup_fns, but it was just fragile and unncecessarily
complicated.

Some comments about ProxyClient cleanup code are updated for clarity here but
no changes are made.

ryanofsky commented:
Rebased 0b45cb5662798553bc51979125d5c549d05e3262 -> 4c3373459d3a2abfa2c5e143f271ec8a4e25b346 ([`pr/eventlock.6`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.6) -> [`pr/eventlock.7`](https://github.com/ryanofsky/libmultiprocess/commits/pr/eventlock.7), [compare](https://github.com/ryanofsky/libmultiprocess/compare/pr/eventlock.6-rebase..pr/eventlock.7)) due to conflict with #181, also adding 3 commits to fix #182, and making minor cleanups to earlier commits (mainly comments)

ryanofsky commented on include/mp/proxy.h:
```diff
@@ -47,6 +48,26 @@ inline void CleanupRun(CleanupList& fns) {
+{
+public:
+    explicit EventLoopRef(EventLoop& loop, std::unique_lock<std::mutex>* lock = nullptr);
+    EventLoopRef(EventLoopRef&& other) noexcept : m_loop(other.m_loop) { other.m_loop = nullptr; }
```
re: https://github.com/bitcoin-core/libmultiprocess/pull/160#discussion_r2149891848

In commit "proxy-io.h: Add EventLoopRef RAII class handle addClient/removeClient refcounting" (9aaeec3678d3b5dad61d2142c37bac18aa6257e0)

> As far as I can tell it is correct to not copy over the lock from the moved object here as long as the moved object has no lock. Should that be asserted?

It's kind of a corner case, and I think either behavior could be useful here. The point of this constructor is to be able to efficiently move a reference count from one owner to another would pointlessly incrementing & decrement the count when creating one reference and dropping the other.

I feel like it *probably* would not be useful to copy the `m_lock` pointer when doing that because practical effect of copying would be to make the new EventLoopRef object destructor skip locking when it is destroyed (it would assert m_lock is locked instead).

On the other hand the default move-constructor would copy the m_lock value so maybe it is less surprising to follow the default behavior. And if the caller really wanted `m_lock` to be null after moving they could set it to null themselves.

So this is a good catch and I think I'd now lean towards adding `m_lock{other.m_lock}` but I don't think the difference should matter in practice.

More broadly I was also thinking about whether I could split EventLoopRef into a superclass without an `m_lock` member and a subclass with one to be able to get rid of the `relock` option added in ea38392960e1edbdb423bebc5cc0a665fa83a253 which requires turning off thread safety analysis in the reset function. At the moment this class seems good at preventing reference counting bugs because it guarantees reference counts will always be decremented if they are incremented, but it doesn't have very easy to understand locking semantics so that's probably an area for improvement.

ryanofsky commented on src/mp/proxy.cpp:
```diff
@@ -270,24 +280,21 @@ void EventLoop::post(kj::Function<void()> fn)
-            while (!done()) {
-                if (!m_async_fns.empty()) {
+            while (m_async_fns) {
+                if (!m_async_fns->empty()) {
```
re: https://github.com/bitcoin-core/libmultiprocess/pull/160#discussion_r2150736150

In commit "Prevent EventLoop async cleanup thread early exit during shutdown" (ea38392960e1edbdb423bebc5cc0a665fa83a253)

> Is there a chance of encountering a race condition here between checking `m_async_fns` and calling `->empty()`?

There shouldn't be just because m_async_fns is guarded by m_mutex which is locked above.

ryanofsky commented:
re: https://github.com/bitcoin-core/libmultiprocess/pull/160#issuecomment-2979687567

> For what it's worth I think the code here could be merged. Further improving exception handling does not all have to happen in here. Getting the EventLoop locking cleanups in is a clear improvement on the baseline in my eyes.

Yes that makes a lot of sense. There are more improvements to be made but I don't think it's good for this PR to just keep growing forever because the next set of improvements depend on current ones.

So I'll plan to not make any more changes here and merge in the next day or so, but I would appreciate if you or Sjors or other reviewers could look at this a little more and add acks if appropriate before then. As mentioned https://github.com/bitcoin-core/libmultiprocess/pull/172#issuecomment-2886589061 I would really like to get away from the practice of merging PRs here without current acks.

sedited reviewed (approved):
ACK 84cf56a0b5f496b02e761dc8fb243cd6a69b3888
